consensus/polybft/state.go

```diff
@@ -256,6 +335,79 @@ func (s *State) list() ([]*StateSyncEvent, error) {
 	return events, nil
 }
 
+// insertStateSyncEvent inserts a new state sync event to state event bucket in db
+func (s *State) insertExitEvent(event *ExitEvent) error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		raw, err := json.Marshal(event)
+		if err != nil {
+			return err
+		}
+
+		bucket := tx.Bucket(exitEventsBucket)
+
+		return bucket.Put(bytes.Join([][]byte{itob(event.Epoch), itob(event.ID), itob(event.BlockNumber)}, nil), raw)
+	})
+}
+
+// getExitEvent returns exit event with given id, which happened in given epoch and given block number
+func (s *State) getExitEvent(exitEventID, epoch, checkpointBlockNumber uint64) (*ExitEvent, error) {
+	var exitEvent *ExitEvent
+
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(exitEventsBucket)
+
+		key := bytes.Join([][]byte{itob(epoch), itob(exitEventID), itob(checkpointBlockNumber)}, nil)
+		v := bucket.Get(key)
+		if v == nil {
+			return &ExitEventNotFoundError{exitEventID, checkpointBlockNumber, epoch}
+		}
+
+		return json.Unmarshal(v, &exitEvent)
+	})
+
+	return exitEvent, err
+}
+
+// getExitEventsByEpoch returns all exit events that happened in the given epoch
+func (s *State) getExitEventsByEpoch(epoch uint64) ([]*ExitEvent, error) {
+	return s.getExitEvents(epoch, func(exitEvent *ExitEvent) bool {
+		return exitEvent.Epoch == epoch
+	})
+}
+
+// getExitEventsForProof returns all exit events that happened in and prior to the given checkpoint block number
+// with respect to the epoch in which block is added
+func (s *State) getExitEventsForProof(epoch, checkpointBlock uint64) ([]*ExitEvent, error) {
```

Comment:
Do you find `getExitEventsUntilBlockNumber` more appropriate perhaps?

Reply:
I would leave it as is. **Until** is not quite right, and it isn't used anywhere but to build proof.

Reply:
Function name should contain what function **does** and not where it is invoked. But as you wish :stuck_out_tongue: